### DISCUSSION TITLE
React native web error: __vite_ssr_import_0__.default is not a function

### DISCUSSION THREAD

**Discussion by zoriya** [vikejs/vike#1637] [Discussion]
### Description

To reproduce:

Simply clone the repro from: https://github.com/zoriya/vike-react-native-web-bug 
OR: 
- Create a new vike app (I used ` yarn create @batijs/app --react`)
 - Install react-native, react-native-web & vite-plugin-react-native-web
 - Add reactNativeWeb to the vite.config.ts (like in the sample below), the project now works if `ssr: false` is set in the +config

```typescript
import ssr from "vike/plugin";
import react from "@vitejs/plugin-react";
import reactNativeWeb from "vite-plugin-react-native-web";
import { defineConfig } from "vite";

export default defineConfig({
	plugins: [react({}), ssr({}), reactNativeWeb()],
});
```

If ssr is enabled (`ssr: true` in +config), the following error happens (please note that using `ssr.noExternal` or `vite-plugin-cjs-interop` does not fix the issue):

```
3:01:35 PM [vike][request(1)] HTTP request: /
3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/modules/prefixStyles/index.js:
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/exports/StyleSheet/compiler/index.js: failed to import "/node_modules/react-native-web/dist/modules/prefixStyles/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/exports/StyleSheet/index.js: failed to import "/node_modules/react-native-web/dist/exports/StyleSheet/compiler/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/modules/createDOMProps/index.js: failed to import "/node_modules/react-native-web/dist/exports/StyleSheet/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/exports/createElement/index.js: failed to import "/node_modules/react-native-web/dist/modules/createDOMProps/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /node_modules/react-native-web/dist/index.js: failed to import "/node_modules/react-native-web/dist/exports/createElement/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module /pages/index/+Page.tsx: failed to import "/node_modules/react-native-web/dist/index.js"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vite] Error when evaluating SSR module virtual:vike:pageConfigValuesAll:server:/pages/index: failed to import "/pages/index/+Page.tsx"
|- TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

3:01:35 PM [vike][request(1)][Error] An error was thrown:
TypeError: __vite_ssr_import_0__.default is not a function
    at eval (/home/zoriya/projects/bati-vike/node_modules/react-native-web/dist/modules/prefixStyles/index.js:7:46)
    at instantiateModule (file:///home/zoriya/projects/bati-vike/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)
3:01:35 PM [vike][request(1)] HTTP response / 500
```

**Comment by brillout** [comment]
I don't think this is related to Vike, but let me know if you have evidence that links this to Vike.

**Reply by zoriya** [reply]
This is a SSR only bug, if it's not vike fault where should I report this?

**Reply by brillout** [reply]
I’d report this to the React Native team.

**Reply by zoriya** [reply]
This is not a react native bug, this works using vite without ssr or using next.

**Reply by brillout** [reply]
Yes, it seems to be an issue between React Native and Vite, but the culprit is usually on the tool’s side, see explanation at https://vike.dev/broken-npm-package.

You can try to reproduce with Vite only (without Vike) by using https://github.com/bluwy/create-vite-extra. That way you show that the issue will occur on all Vite-based SSR frameworks.

**Reply by brillout** [reply]
Also FYI 3eb918eb708c04919d6dc76336eb8b65d9fa2ac7.

**Reply by brillout** [reply]
Making progress: https://github.com/brillout/vike-react-native-web-bug but still getting an error:

```
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
8:15:48 PM [vike][request(1)] Following error was thrown by the onRenderHtml() hook defined at vike-react/renderer/onRenderHtml
TypeError: Cannot read properties of null (reading 'useContext')
    at Object.useContext6 (/home/rom/tmp/vike-react-native-web-bug/node_modules/.vite/deps_ssr/react-native-web.js?v=97d047a8:3119:29)
```

It would be great if someone reaches out to the RN team about this. This affects all Vite-based frameworks.

**Reply by zoriya** [reply]
Thanks for looking into this! I could reproduce it with vite only (https://stackblitz.com/edit/rnw-ssr-vite) but I could not get as much logs as with vike (I'm probably missing a flag).

React native web is not maintained by the RN team but by necolas, a twitter employee (last time I checked at least).
I don't think we will get much traction from this side (https://github.com/necolas/react-native-web/issues/2446 or https://github.com/necolas/react-native-web/issues/2200).

**Reply by brillout** [reply]
Actually, it seems like necolas isn't aware that his project doesn't work on any Vite-based SSR framework (Remix, Vike, Astro, etc.). So I'd actually very much reach out to him with your minimal Vite reproduction.

**Reply by brillout** [reply]
@zoriya Keep me updated, it would be gread to eventually create a `vike.dev/react-native` (currently 404) page.

**Reply by zoriya** [reply]
I'll create issues on both react-native-web and vite and link it here. I do think that react-native+vike has great potential, today next+expo is becoming pretty popular (just looking at https://github.com/nandorojo/solito) and I would like to get away from next.

**Reply by zoriya** [reply]
https://github.com/vitejs/vite/issues/16679

**Reply by brillout** [reply]
I recommend to also reach out to the react native web maintainer.

**Reply by zoriya** [reply]
https://github.com/necolas/react-native-web/issues/2668

**Reply by zoriya** [reply]
I added `inline-style-prefixer` to the list of csInterop dependencies on the vanilla vite repro and got a different error than with vike (an import error instead of a hook error). You can see logs on the rnw issue above and here is the repro: https://github.com/zoriya/vite-rnw-ssr-repro

**Reply by brillout** [reply]
The error you're getting can be fixed with https://github.com/brillout/vike-react-native-web-bug/blob/953670a90f46a89511c004d5b9f13fa850a09091/vite.config.ts#L14-L18 and you'll likely get the same hook error then.

**Reply by zoriya** [reply]
Thanks, I do get the same issue now.

**Comment by zoriya** [comment]
Is there a go-to way of fixing esmodules upstreams? For example, `@expo/html-elements` ships esm code but do not use extensions in export/imports. (as in `import Text from "./Text";` instead of `from "./Text.js"`). 

I could not find a way to ignore this locally (via a vite option or using bun) and I don't know what's the way to fix upstream. Note that `inline-style-prefixer` has the same issue in its esm build (see https://github.com/robinweser/inline-style-prefixer/issues/232).

**Reply by brillout** [reply]
Maybe you can try to pre-bundle them with Wepack? In theory that should fix the situation for good, but you may need to dedupe React so that could be tricky as well.

**Reply by brillout** [reply]
Or with esbulid (much faster) if it works.
